qiskit_nature/algorithms/initial_points/mp2_initial_point.py

```diff
@@ -0,0 +1,261 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The MP2InitialPoint class to compute an initial point for the VQE Ansatz parameters."""
+
+from __future__ import annotations
+
+import numpy as np
+
+from qiskit_nature.properties.second_quantization.electronic import ParticleNumber, ElectronicEnergy
+from qiskit_nature.properties.second_quantization.electronic.bases import ElectronicBasis
+from qiskit_nature.circuit.library import UCC
+from qiskit_nature.properties.second_quantization.second_quantized_property import (
+    GroupedSecondQuantizedProperty,
+)
+
+from .initial_point import InitialPoint
+
+
+class MP2InitialPoint(InitialPoint):
+    """Moller-Plesset 2nd order initial point generator.
+
+    A class for computing the Moller-Plesset 2nd order (MP2) corrections to use
+    as an initial point for VQE alongside a UCC ansatz.
+
+    The MP2 calculation requires ``orbital_energies`` from the ``ElectronicEnergy``
+    in the driver result.
+
+    An MP2InitialPoint uses the same excitations as generated by UCC to
+    ensure that the coefficients are mapped correctly.
+
+    The coefficients and energy deltas are computed when the object is
+    initialized. The ``initial_point`` property may then be used with VQE.
+
+    Only initial point values that correspond to double-excitations will be
+    non-zero.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.threshold = 1e-12
+        self._corrections: dict[str, tuple[float, float]] = None
+
+    @property
+    def driver_result(self) -> GroupedSecondQuantizedProperty:
+        """The driver result."""
+        return self._driver_result
+
+    @driver_result.setter
+    def driver_result(self, result: GroupedSecondQuantizedProperty) -> None:
+        """The driver result."""
+
+        electronic_energy = result.get_property(ElectronicEnergy)
+        self.electronic_energy = electronic_energy
+
+        particle_number = result.get_property(ParticleNumber)
+        self.num_spin_orbitals = particle_number.num_spin_orbitals
```

Comment:
@mrossinek that's a very nice suggestion. I've made this change. I guess at this point, we could just pass in ElectronicEnergy rather than the grouped_property, but I've kept it to grouped_property so a user can just pass in a driver result and with an eye to other initial points and might need different properties.